### PR TITLE
TLSv1.3: Export keying material

### LINKED ISSUE #3680: TLSv1.3 Exporters
https://tools.ietf.org/html/draft-ietf-tls-tls13-20#section-7.5

TLSv1.3 defines exporters which use different construction from the original RFC 5705.
OpenSSL has  SSL_export_keying_material, but it seems that , even if TLSv1.3 is negotiated, the underlying export function is still tls1_export_keying_material, which uses PRF version.

### PR DESCRIPTION
**Pull request by mattcaswell**
This updates SSL_export_keying_material() for TLSv1.3. I've added a test, but there are no test vectors available, so all it does is check that both sides create the same value and that it looks sane.

Fixes #3680

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: doc/man3/SSL_export_keying_material.pod [hunk 1]
```diff
@@ -32,8 +33,10 @@ pointed to by B<context> and should be B<contextlen> bytes long. Provision of
 a context is optional. If the context should be omitted entirely then
 B<use_context> should be set to 0. Otherwise it should be any other value. If
 B<use_context> is 0 then the values of B<context> and B<contextlen> are ignored.
-Note that a zero length context is treated differently to no context at all, and
-will result in different keying material being returned.
+Note that in TLSv1.2 and below a zero length context is treated differently to
```
Comment (richsalz) [inline doc/man3/SSL_export_keying_material.pod]: "differently from"

File: ssl/tls13_enc.c [hunk 2]
```diff
@@ -587,3 +597,41 @@ int tls13_alert_code(int code)
...
+    EVP_MD_CTX *ctx = EVP_MD_CTX_new();
+    unsigned int hashsize;
+    int ret = 0;
+
+    if (ctx == NULL)
+        goto err;
+
+    if (!SSL_is_init_finished(s))
+        goto err;
+
```
Comment (richsalz) [inline ssl/tls13_enc.c]: could merge these into one if test.  if not, at least remove the blank line.

File: test/sslapitest.c [hunk 3]
```diff
@@ -2486,6 +2486,145 @@ static int test_serverinfo(int tst)
...
+    if (tst == 3)
+        return 1;
+#endif
+    if (!TEST_true(create_ssl_ctx_pair(TLS_server_method(),
+                                       TLS_client_method(), &sctx,
+                                       &cctx, cert, privkey)))
+        goto end;
+
+    switch (tst) {
+    case 0:
```
Comment (richsalz) [inline test/sslapitest.c]: why not use a table lookup?  ```proto = protocols[tst]```

File: test/sslapitest.c [hunk 4]
```diff
@@ -2486,6 +2486,145 @@ static int test_serverinfo(int tst)
...
+               /* Different contexts should produce different results */
+            || !TEST_mem_ne(ckeymat1, sizeof(ckeymat1), ckeymat2,
+                            sizeof(ckeymat2)))
+        goto end;
+
+    /*
+     * Check that an empty context and no context produce different results in
+     * protocols less than TLSv1.3. In TLSv1.3 they should be the same.
+     */
+    if ((tst != 3 && !TEST_mem_ne(ckeymat2, sizeof(ckeymat2), ckeymat3,
```
Comment (richsalz) [inline test/sslapitest.c]: put the linebreak after the ```&&``` if the TEST_mem line would fit on one line.
same commen t for the following expression, too.

File: doc/man3/SSL_export_keying_material.pod (hunk 1 above)
Comment (mattcaswell) [inline doc/man3/SSL_export_keying_material.pod]: Are we going to have another British vs US English discussion ;-)? But "differently from" is equally acceptable in British English, so I've changed it.

File: ssl/tls13_enc.c (hunk 2 above)
Comment (mattcaswell) [inline ssl/tls13_enc.c]: Done

File: test/sslapitest.c (hunk 3 above)
Comment (mattcaswell) [inline test/sslapitest.c]: Good idea. Done.

File: test/sslapitest.c (hunk 4 above)
Comment (mattcaswell) [inline test/sslapitest.c]: It didn't help. No change made.

**Comment by mattcaswell** [comment]
New commit pushed addressing feedback.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**